gear-test/src/manager.rs

```diff
@@ -79,66 +85,70 @@ impl CollectState for InMemoryExtManager {
+                self.move_waiting_msgs_to_queue(program_id);
                 false
             }
         };
     }
     fn gas_burned(&mut self, _message_id: MessageId, _origin: ProgramId, _amount: u64) {}
     fn message_consumed(&mut self, message_id: MessageId) {
         if let Some(index) = self
-            .message_queue
+            .dispatch_queue
             .iter()
-            .position(|msg| msg.id() == message_id)
+            .position(|d| d.message.id() == message_id)
         {
-            self.message_queue.remove(index);
+            self.dispatch_queue.remove(index);
         }
     }
-    fn send_message(&mut self, _message_id: MessageId, message: Message) {
-        let id = message.dest();
-        if self.programs.borrow().contains_key(&id) {
-            let mut borrowed_list = self.waiting_init.borrow_mut();
-            if let (None, Some(list)) = (message.reply(), borrowed_list.get_mut(&id)) {
-                list.push(message.id);
-                self.wait_list.insert((id, message.id), message);
+    fn send_dispatch(&mut self, _message_id: MessageId, dispatch: Dispatch) {
+        let dest = dispatch.message.dest();
+        if self.programs.borrow().contains_key(&dest) {
+            if let (DispatchKind::Handle, Some(list)) =
+                (dispatch.kind, self.waiting_init.borrow_mut().get_mut(&dest))
+            {
+                let message_id = dispatch.message.id();
+                list.push(message_id);
+                self.wait_list.insert((dest, message_id), dispatch);
             } else {
-                self.message_queue.push_back(message);
+                self.dispatch_queue.push_back(dispatch);
             }
         } else {
-            self.log.push(message);
+            self.log.push(dispatch.message);
         }
     }
     fn wait_dispatch(&mut self, dispatch: Dispatch) {
         self.message_consumed(dispatch.message.id());
-        self.wait_list.insert(
-            (dispatch.message.dest(), dispatch.message.id()),
-            dispatch.message,
-        );
+        self.wait_list
+            .insert((dispatch.message.dest(), dispatch.message.id()), dispatch);
     }
     fn wake_message(
         &mut self,
         _message_id: MessageId,
         program_id: ProgramId,
         awakening_id: MessageId,
     ) {
-        if let Some(msg) = self.wait_list.remove(&(program_id, awakening_id)) {
-            self.message_queue.push_back(msg);
+        if let Some(dispatch) = self.wait_list.remove(&(program_id, awakening_id)) {
+            self.dispatch_queue.push_back(dispatch);
         }
     }
     fn update_nonce(&mut self, program_id: ProgramId, nonce: u64) {
-        if let Some(prog) = self.programs.borrow_mut().get_mut(&program_id) {
+        let mut programs = self.programs.borrow_mut();
+        if let Some(prog) = programs
+            .get_mut(&program_id)
+            .expect("Program not found in storage")
```

Comment:
this expect panic for each case, except program is Active and it's bad. Terminated program and "Program not found" (user meant) is different cases